### PR TITLE
[Merged by Bors] - Delete orphaned cluster resources

### LINKED ISSUE #151: Handle stale information/clean up stale resources
UPDATE: During the daily, it was mentioned that if the individual Service objects for each Pod have the owner reference set, then Kubernetes will clean these Services automatically when the Pods are deleted. It was decided that this issue **is not blocked**.

~~**Blocked by**: https://github.com/stackabletech/hdfs-operator/issues/174 - it makes sense to first NodePort handling to the listener operator before implementing this issue.~~

Currently our operators will not act on removed information from the CR in some/most/all cases.

One example:
HBase operator has three roles (master, regionServer, restServer). If I create a HBase server CR with a restServer component and then remove it later (entirely, not setting replicas to 0) our operator will not clean up the STS that belongs to this role.

## Proposed solution

Use the `ClusterResource` struct from operator-rs to manage Kubernetes resources belonging to a Cluster object. An example of its usage can be found in the Superset Operator: https://github.com/stackabletech/superset-operator/blob/main/rust/operator-binary/src/superset_controller.rs#L241

## Acceptance criteria

- [x] This is done when all stale Kubernetes resources are cleaned up. A resource becomes stale when it's not part of the current cluster definition anymore.
- [x] `ZNode` resources are not deleted because the operator cannot know if they are stale or not.
- [x] `PersistentVolumeClaims` used by the cluster are never deleted by the operator.
- [x] There is at least one test.
- [ ] Documentation on operator implementation is updated with information regarding handling of stale resources.
- [x] Upgrade to the latest version of` operator-rs` (`0.25` at the moment).

NOTE: This is part of an epic (https://github.com/stackabletech/issues/issues/203) and _might_ not apply to this operator. If that is the case please comment on this issue and just close it. This issue was created as part of a special bulk creation operation.

### PR DESCRIPTION
**Pull request by maltesander**
# Description

- Added orphaned cluster resources for StatefulSet, ConfigMap, Discovery ConfigMap and Rolegroup Service
- Added owner refs from the pods  to pod services (pod_svc_controller)
- Added owner refs for serviceaccount and rolebinding (named {cluster-name}-serviceaccount and {cluster-name}-rolebinding) which are deleted upon deleting the cluster

fixes https://github.com/stackabletech/hdfs-operator/issues/151

test: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/17/

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator/src/hdfs_controller.rs [hunk 1]
```diff
@@ -60,11 +61,18 @@ pub async fn reconcile_hdfs(hdfs: Arc<HdfsCluster>, ctx: Arc<Ctx>) -> HdfsOperat
...
+    )
+    .map_err(|e| Error::CreateClusterResources { source: e })?;
+
     let discovery_cm = build_discovery_configmap(&hdfs, &namenode_podrefs)
         .map_err(|e| Error::BuildDiscoveryConfigMap { source: e })?;
 
-    client
-        .apply_patch(FIELD_MANAGER_SCOPE, &discovery_cm, &discovery_cm)
+    cluster_resources
+        .add(client, &discovery_cm)
```
Comment (razvan) [inline rust/operator/src/hdfs_controller.rs]: As discussed, this is dependant on the cluster definition and thus can never become an orphan (or even stale).

**Review by razvan — APPROVED** [review]
Successfully tested with OpenShift and nicely done.

File: rust/operator/src/hdfs_controller.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator/src/hdfs_controller.rs]: done

**Review by razvan — APPROVED** [review]
lgtm

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**